### PR TITLE
Implement user-specific code claiming

### PR DESCRIPTION
**Pull request by lowjoel**
Also implements:
 - Inviting users by manually keying names/emails
 - Users to claim their own user-specific course registration code

What's left:
 - Sending emails to users notifying them
 - Checking that users can register for multiple instances
 - Checking that users can claim their accounts after being invited.

### CODE REVIEW

**Comment by lowjoel** [comment]
@allenwq any comments?

**Comment by allenwq** [comment]
@lowjoel Yah, I'm still reviewing this.

**Comment by lowjoel** [comment]
sorry, stuff got big quick

File: app/models/course/registration.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+class Course::Registration
+  extend ActiveModel::Model
```
Comment (allenwq) [inline app/models/course/registration.rb]: Should be `include` ?

File: app/services/course/user_invitation_service.rb [hunk 2]
```diff
@@ -57,6 +57,18 @@ def load_from_file(file)
...
 
+  # Invites the users from the form submission, which reflects the actual model associations.
+  #
+  # We do not use this format in the service object because it is very clumsy.
+  #
+  # @param [Hash] users The attributes from the client.
+  # @return [void]
+  def invite_from_form(users)
+    invite_users(users.map do |(_, value)|
+      { name: value[:course_user][:name], email: value[:user_email][:email] }
```
Comment (allenwq) [inline app/services/course/user_invitation_service.rb]: Shall we validate the email addresses inputed by user ?

Comment (lowjoel) [inline app/services/course/user_invitation_service.rb]: done by the user email model.

File: config/routes.rb [hunk 3]
```diff
@@ -97,11 +97,11 @@
       get 'components' => 'settings#components'
       patch 'components' => 'settings#update_components'
 
-      resources :users, except: [:index, :show, :new, :edit] do
+      resources :users, except: [:index, :show, :new, :create, :edit] do
```
Comment (allenwq) [inline config/routes.rb]: `only: [:update]` ?

Comment (lowjoel) [inline config/routes.rb]: done

File: app/services/course/user_invitation_service.rb (hunk 2 above)
Comment (allenwq) [inline app/services/course/user_invitation_service.rb]: And will simple_form show the correct errors when email is invalid ?

Comment (lowjoel) [inline app/services/course/user_invitation_service.rb]: you can try. I want to deal with edge cases later, this PR is big as it stands

File: app/models/course/registration.rb (hunk 1 above)
Comment (allenwq) [inline app/models/course/registration.rb]: http://guides.rubyonrails.org/active_model_basics.html#model

File: app/services/course/user_invitation_service.rb (hunk 2 above)
Comment (allenwq) [inline app/services/course/user_invitation_service.rb]: No problem, You can fix in separate PR.

File: app/models/course/registration.rb (hunk 1 above)
Comment (lowjoel) [inline app/models/course/registration.rb]: fixed thanks.

Comment (allenwq) [inline app/models/course/registration.rb]: I guess this include is useless since tests passed in both cases  ><

Comment (lowjoel) [inline app/models/course/registration.rb]: not necessarily...

**Pull request merged**